File: Data Structure/Linked Lists/Single Linked List/SingleLinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None   
    
class SingleLinkedList :
    def __init__(self):
       self.head = None


    def search(self, target):
        temp = self.head

        while(temp!=None):
            if(temp.data == target):
                return True
            
            temp = temp.next
        
        return False
    
    
    def traverse(self):
        temp = self.head
        while(temp!=None):
            print(temp.data)
            temp = temp.next
        
    
    
    def length (self):
        length = 0
        temp = self.head
        while (temp!=None):
            length+=1
            temp=temp.next
        return length
    
    
    def insertionAtBeginning(self, data):
        temp = Node(data)
        temp.next = self.head
        self.head = temp
    
    
    def insertionAtEnd(self, data):
        temp = Node(data)
        head_temp = self.head
        if (self.head == None):
            self.head = temp
        else:
            while(head_temp.next != None):
                head_temp = head_temp.next
            
            head_temp.next = temp
        
    
    
    def insertionAtPosition(self, position, data):
        temp = Node(data)
        headTemp = self.head
        if(position == 0 or headTemp == None):
            temp.next = headTemp
            headTemp = temp
            return
        
        for _ in range(position):
            if(headTemp.next != None):
                headTemp = headTemp.next
        
        temp.next = headTemp.next
        headTemp.next = temp
    
    
    def deletionAtBeginning(self):
        if(self.head == None):
            print("No node")
            return
        
        self.head = self.head.next
    
    
    def deletionAtEnd(self):
        if(self.head == None):
            print("No node")
            return

        if(self.head.next == None):
            self.head = None
            return
    
        temp = self.head
    
        while(temp.next.next != None):
            temp = temp.next
        
        temp.next = None
    
    
    def deletionAtPosition(self, position):
        if(self.head == None):
            print("No node")
            return
        
    
        if(position==0):
            self.head = self.head.next
            return
        
    
        temp = self.head
    
        for _ in range(position):
            if(temp.next != None):
                temp = temp.next
        
    
        if(temp.next != None):
            temp.next = temp.next.next
        else:
            print("No node")
            return
        
    


if __name__ == '__main__':
    sll = SingleLinkedList()
    sll.insertionAtEnd("B")
    sll.insertionAtEnd("C")
    sll.insertionAtEnd("D")
    sll.traverse()
    print(sll.search("D"))
    print("Length", sll.length())
    sll.insertionAtBeginning("A")
    sll.insertionAtEnd("F")
    sll.insertionAtPosition( 3, "E")
    sll.deletionAtBeginning()
    sll.deletionAtEnd()
    sll.deletionAtPosition(1)
    sll.traverse()
    

File: Data Structures/Linked Lists/Single Linked List/SingleLinkedList.cpp
#include<iostream>
using namespace std;

class SingleLinkedList{
    public:
    string data;
    SingleLinkedList* next;

    public:
    SingleLinkedList(string data1){
        data = data1;
        next = nullptr;
    }
};

bool search(SingleLinkedList* node, string target){
    while(node->next != nullptr){
        if(node->data == target){
            return true;
        }
        node = node->next;
    }
    return false;
}

void traverse(SingleLinkedList* node){
    while(node!=nullptr){
        cout<< node->data<<endl;
        node = node->next;
    }
}

int length (SingleLinkedList* node){
    int temp = 0;
    while (node->next != nullptr)
    {
        temp+=1;
        node=node->next;
    }
    return temp;
}

void insertionAtBeginning(SingleLinkedList** head, string data){
    SingleLinkedList* temp = new SingleLinkedList(data);
    temp->next = *head;
    *head = temp;
}

void insertionAtEnd(SingleLinkedList** Node, string data){
    SingleLinkedList* temp = new SingleLinkedList(data);
    SingleLinkedList* head_temp = *Node;
    if ((*Node) == nullptr){
        (*Node) = temp;
    }
    else{
        while(head_temp->next != nullptr){
            head_temp = head_temp->next;
        }
        head_temp->next = temp;
    }
}

void insertionAtPosition(SingleLinkedList** Node, int position, string data){
    SingleLinkedList* temp = new SingleLinkedList(data);
    SingleLinkedList* headTemp = *Node;
    if(position == 0 || headTemp == nullptr){
        temp->next = headTemp;
        headTemp = temp;
        return;
    }

    for(int i = 0; i < position && headTemp->next != nullptr; i++){
        headTemp = headTemp->next;
    }
    temp->next = headTemp->next;
    headTemp->next = temp;
}

void deletionAtBeginning(SingleLinkedList** Node){
    if(*Node == nullptr){
        cout << "No node" << endl;
        return;
    }
    SingleLinkedList* temp = *Node;
    *Node = (*Node)->next;
    delete temp;
}

void deletionAtEnd(SingleLinkedList** Node){
    if(*Node == nullptr){
        cout << "No node" << endl;
        return;
    }

    if((*Node)->next == nullptr){
        delete *Node;
        *Node = nullptr;
        return;
    }

    SingleLinkedList* temp = *Node;

    while(temp->next->next != nullptr){
        temp = temp->next;
    }

    delete temp->next;
    temp->next = nullptr;
}

void deletionAtPosition(SingleLinkedList** Node, int position){
    if(*Node == nullptr){
        cout<<"No node";
        return;
    }

    if(position==0){
        SingleLinkedList * temp = *Node;
        *Node = (*Node)->next;
        delete temp;
        return;
    }

    SingleLinkedList * temp = *Node;

    for(int i = 1; i<position && temp->next != nullptr; i++){
        temp = temp->next;
    }

    if(temp->next != nullptr){
        SingleLinkedList * temp2 = temp->next;
        temp->next = temp->next->next;
        delete temp2;
    }
    else{
        cout << "No node";
        return;
    }
}

int main(){
    SingleLinkedList* a =  new SingleLinkedList("B");
    SingleLinkedList* b =  new SingleLinkedList("C");
    SingleLinkedList* c =  new SingleLinkedList("D");
    a->next = b;
    b->next = c;
    traverse(a);
    cout<<search(a,"D")<<endl;
    cout<<"Length: "<<length(a)<<endl;
    insertionAtBeginning(&a,"A");
    insertionAtEnd(&a,"F");
    insertionAtPosition(&a, 3, "E");
    // deletionAtBeginning(&a);
    // deletionAtEnd(&a);
    deletionAtPosition(&a, 1);
    traverse(a);
    return 0;
}